cmd/juju/application/deployer/deployer.go

```diff
@@ -0,0 +1,546 @@
+}
+
+// NewDeployerFactory returns a factory setup with the API and
+// function dependencies required by every deployer.
+func NewDeployerFactory(dep DeployerDependencies) DeployerFactory {
+	d := &factory{
+		clock:                jujuclock.WallClock,
+		model:                dep.Model,
+		newConsumeDetailsAPI: dep.NewConsumeDetailsAPI,
+		steps:                dep.Steps,
+	}
+	if dep.DeployResources == nil {
+		d.deployResources = resourceadapters.DeployResources
+	}
+	return d
+}
+
+// Deployer defines the functionality of a deployer returned by the
+// factory.
+type Deployer interface {
+	// PrepareAndDeploy finishes preparing to deploy a charm or bundle,
+	// then deploys it.  This is done as one step to accomidate the
+	// call being wrapped by block.ProcessBlockedError.
+	PrepareAndDeploy(*cmd.Context, DeployerAPI, *store.CharmStoreAdaptor) error
+}
+
+// GetDeployer returns the correct deployer to use based on the cfg provided.
+// A ModelConfigGetter and CharmStoreAdaptor needed to find the deployer.
+func (d *factory) GetDeployer(cfg DeployerConfig, getter ModelConfigGetter, cstoreAPI *store.CharmStoreAdaptor) (Deployer, error) {
+	d.setConfig(cfg)
+	maybeDeployers := []func() (Deployer, error){
+		d.maybeReadLocalBundle,
+		func() (Deployer, error) { return d.maybeReadLocalCharm(getter) },
+		d.maybePredeployedLocalCharm,
+		func() (Deployer, error) { return d.maybeReadCharmstoreBundle(cstoreAPI) },
+		d.charmStoreCharm, // This always returns a Deployer
+	}
+	for _, d := range maybeDeployers {
+		if deploy, err := d(); err != nil {
+			return nil, errors.Trace(err)
+		} else if deploy != nil {
+			return deploy, nil
+		}
+	}
+	return nil, errors.NotFoundf("suitable Deployer")
+}
+
+func (d *factory) setConfig(cfg DeployerConfig) {
+	d.placementSpec = cfg.PlacementSpec
+	d.placement = cfg.Placement
+	d.numUnits = cfg.NumUnits
+	d.attachStorage = cfg.AttachStorage
+	d.charmOrBundle = cfg.CharmOrBundle
+	d.bundleOverlayFile = cfg.BundleOverlayFile
+	d.channel = cfg.Channel
+	d.series = cfg.Series
+	d.force = cfg.Force
+	d.dryRun = cfg.DryRun
+	d.applicationName = cfg.ApplicationName
+	d.configOptions = cfg.ConfigOptions
+	d.constraints = cfg.Constraints
+	d.storage = cfg.Storage
+	d.bundleStorage = cfg.BundleStorage
+	d.devices = cfg.Devices
+	d.bundleDevices = cfg.BundleDevices
+	d.resources = cfg.Resources
+	d.bindings = cfg.Bindings
+	d.useExisting = cfg.UseExisting
+	d.bundleMachines = cfg.BundleMachines
+	d.trust = cfg.Trust
+	d.flagSet = cfg.FlagSet
+}
+
+// DeployerDependencies are required for any deployer to be run.
+type DeployerDependencies struct {
+	DeployResources      resourceadapters.DeployResourcesFunc
+	Model                ModelCommand
+	NewConsumeDetailsAPI func(url *charm.OfferURL) (ConsumeDetails, error)
+	Steps                []DeployStep
+}
+
+// DeployerConfig is the data required to choose a deployer and then run
+// PrepareAndDeploy.
+type DeployerConfig struct {
```

Comment:
This is terrible, but I don't see how to fix this in the short term.

Can we add a TODO to refactor this config into a better structure?

Reply:
adding